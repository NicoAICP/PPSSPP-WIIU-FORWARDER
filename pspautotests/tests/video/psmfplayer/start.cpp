#include "shared.h"
#include <pspthreadman.h>
#include <pspdisplay.h>

void testStart(const char *title, SceUID *psmf, int vcodec = 0x0e, int vnum = 0, int acodec = 0x0f, int anum = 0, int mode = 0, int speed = 1, int startpts = 0) {
	PsmfPlayerData data = {vcodec, vnum, acodec, anum, mode, speed};
	int result = scePsmfPlayerStart(psmf, &data, startpts);
	scePsmfPlayerGetCurrentPlayMode(psmf, &mode, &speed);
	scePsmfPlayerGetCurrentAudioStream(psmf, &acodec, &anum);
	scePsmfPlayerGetCurrentVideoStream(psmf, &vcodec, &vnum);
	if (result == 0) {
		checkpoint("%s: OK (status=%x)", title, scePsmfPlayerGetCurrentStatus(psmf));

		int pts = -1337;
		int tries = 500;
		u8 audioData[0x2000];
		while (scePsmfPlayerGetCurrentPts(psmf, (u64 *)&pts) != 0 && tries > 0) {
			--tries;
			scePsmfPlayerUpdate(psmf);
			PsmfVideoData videoData = {0, getPsmfPlayerDisplayBuf()};
			scePsmfPlayerGetVideoData(psmf, &videoData);
			scePsmfPlayerGetAudioData(psmf, audioData);
			sceKernelDelayThread(100);
		}
		checkpoint("%s: mode=%d, speed=%d, audio=%d/%d, video=%d/%d, pts=%d", title, scePsmfPlayerGetCurrentStatus(psmf), mode, speed, acodec, anum, vcodec, vnum, pts);
	} else {
		checkpoint("%s: Failed: %08x", title, result);
	}
}

extern "C" int main(int argc, char *argv[]) {
	loadPsmfPlayer();

	sceMpegInit();
	checkpointNext("Init");

	checkpointNext("Players (get):");
	testStart("  Initial", createPsmfPlayerInitial());
	testStart("  Standby", createPsmfPlayerStandby());
	// Crashes.
	//testStart("  NULL", NULL);
	// TODO: Not the most ideal test video...
	SceUID *psmfPlaying = createPsmfPlayerPlaying("host0:/tests/video/psmfplayer/test.pmf");
	testStart("  Playing", psmfPlaying);

	// Force the video ahead a bit.
	for (int j = 0; j < 20; ++j) {
		u8 audioData[0x2000];
		scePsmfPlayerUpdate(psmfPlaying);
		PsmfVideoData videoData = {0, getPsmfPlayerDisplayBuf()};
		scePsmfPlayerGetVideoData(psmfPlaying, &videoData);
		scePsmfPlayerGetAudioData(psmfPlaying, audioData);
		sceKernelDelayThread(100);
	}

	testStart("  Twice", psmfPlaying);
	SceUID psmf2 = *psmfPlaying;
	testStart("  Copy", &psmf2);
	testStart("  Finished", createPsmfPlayerFinished());
	psmf2 = 0;
	testStart("  Zero", &psmf2);

	checkpointNext("Modes:");
	static const int modes[] = {-2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 10, 16, 0x80000000};
	for (size_t i = 0; i < ARRAY_SIZE(modes); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Mode %d", modes[i]);
		testStart(temp, createPsmfPlayerStandby(), 0x0e, 0, 0x0f, 0, modes[i], 1, 0);
	}

	checkpointNext("Speeds:");
	static const int speeds[] = {-2, -1, 0, 1, 2, 3, 4, 5, 10, 16, 0x80000000};
	for (size_t i = 0; i < ARRAY_SIZE(speeds); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Speed %d", speeds[i]);
		testStart(temp, createPsmfPlayerStandby(), 0x0e, 0, 0x0f, 0, 0, speeds[i], 0);
	}
	
	checkpointNext("Start pts:");
	static const int pts[] = {-1000, -1, 0, 1, 1000, 3003, 6006, 87087, 276276, 279278, 279279, 282282, 0x7FFFFFFF};
	for (size_t i = 0; i < ARRAY_SIZE(pts); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Pts %d", pts[i]);
		testStart(temp, createPsmfPlayerStandby(), 0x0e, 0, 0x0f, 0, 0, 1, pts[i]);
	}

	checkpointNext("Video codecs:");
	static const int videoCodecs[] = {0x00, 0x01, 0x0e, 0x0f, 0x1e, 0x8000000e};
	for (size_t i = 0; i < ARRAY_SIZE(videoCodecs); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Video codec %08x", videoCodecs[i]);
		testStart(temp, createPsmfPlayerStandby(), videoCodecs[i], 0, 0x0f, 0, 0, 1, 0);
	}

	checkpointNext("Audio codecs (no audio):");
	static const int audioCodecs[] = {0x00, 0x01, 0x0e, 0x0f, 0x1e, 0x8000000f};
	for (size_t i = 0; i < ARRAY_SIZE(audioCodecs); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Audio codec %08x", audioCodecs[i]);
		testStart(temp, createPsmfPlayerStandby(), 0x0e, 0, audioCodecs[i], 0, 0, 1, 0);
	}

	checkpointNext("Audio codecs (with audio):");
	for (size_t i = 0; i < ARRAY_SIZE(audioCodecs); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  Audio codec %08x", audioCodecs[i]);
		testStart(temp, createPsmfPlayerStandby("host0:/tests/video/psmfplayer/test_streams.pmf"), 0x0e, 0, audioCodecs[i], 0, 0, 1, 0);
	}

	checkpointNext("Streams:");
	testStart("  Video stream -1", createPsmfPlayerStandby(), 0x0e, -1, 0x0f, 0, 0, 1, 0);
	testStart("  Video stream 1", createPsmfPlayerStandby(), 0x0e, 1, 0x0f, 0, 0, 1, 0);
	testStart("  Video stream 1 (exists)", createPsmfPlayerStandby("host0:/tests/video/psmfplayer/test_streams.pmf"), 0x0e, 1, 0x0f, 0, 0, 1, 0);
	testStart("  Video stream 2", createPsmfPlayerStandby("host0:/tests/video/psmfplayer/test_streams.pmf"), 0x0e, 2, 0x0f, 0, 0, 1, 0);
	testStart("  Audio stream -1", createPsmfPlayerStandby(), 0x0e, 0, 0x0f, -1, 0, 1, 0);
	testStart("  Audio stream 1", createPsmfPlayerStandby(), 0x0e, 0, 0x0f, 1, 0, 1, 0);
	testStart("  Audio stream 1 (exists)", createPsmfPlayerStandby("host0:/tests/video/psmfplayer/test_streams.pmf"), 0x0e, 0, 0x0f, 1, 0, 1, 0);
	testStart("  Audio stream 2", createPsmfPlayerStandby("host0:/tests/video/psmfplayer/test_streams.pmf"), 0x0e, 0, 0x0f, 2, 0, 1, 0);

	sceMpegFinish();
	unloadPsmfPlayer();
	return 0;
}