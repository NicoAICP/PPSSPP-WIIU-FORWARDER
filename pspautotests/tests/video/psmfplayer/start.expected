[x] Init

[x] Players (get):
[r]   Initial: Failed: 80616001
[r]   Standby: OK (status=4)
[r]   Standby: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Playing: OK (status=4)
[r]   Playing: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Twice: OK (status=4)
[x]   Twice: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Copy: OK (status=4)
[x]   Copy: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Finished: OK (status=4)
[x]   Finished: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[x]   Zero: Failed: 80616001

[x] Modes:
[r]   Mode -2: Failed: 80616008
[r]   Mode -1: Failed: 80616008
[r]   Mode 0: OK (status=4)
[r]   Mode 0: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Mode 1: OK (status=4)
[r]   Mode 1: mode=4, speed=1, audio=1/-1, video=-1/0, pts=0
[r]   Mode 2: OK (status=4)
[r]   Mode 2: mode=4, speed=2, audio=1/-1, video=-1/0, pts=0
[r]   Mode 3: OK (status=4)
[r]   Mode 3: mode=4, speed=3, audio=1/-1, video=-1/0, pts=0
[r]   Mode 4: Failed: 80616008
[r]   Mode 5: Failed: 80616008
[r]   Mode 6: Failed: 80616008
[r]   Mode 7: Failed: 80616008
[r]   Mode 8: Failed: 80616008
[r]   Mode 10: Failed: 80616008
[r]   Mode 16: Failed: 80616008
[r]   Mode -2147483648: Failed: 80616008

[x] Speeds:
[r]   Speed -2: OK (status=4)
[r]   Speed -2: mode=4, speed=0, audio=-2/-1, video=-1/0, pts=0
[r]   Speed -1: OK (status=4)
[r]   Speed -1: mode=4, speed=0, audio=-1/-1, video=-1/0, pts=0
[r]   Speed 0: OK (status=4)
[r]   Speed 0: mode=4, speed=0, audio=0/-1, video=-1/0, pts=0
[r]   Speed 1: OK (status=4)
[r]   Speed 1: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Speed 2: OK (status=4)
[r]   Speed 2: mode=4, speed=0, audio=2/-1, video=-1/0, pts=0
[r]   Speed 3: OK (status=4)
[r]   Speed 3: mode=4, speed=0, audio=3/-1, video=-1/0, pts=0
[r]   Speed 4: OK (status=4)
[r]   Speed 4: mode=4, speed=0, audio=4/-1, video=-1/0, pts=0
[r]   Speed 5: OK (status=4)
[r]   Speed 5: mode=4, speed=0, audio=5/-1, video=-1/0, pts=0
[r]   Speed 10: OK (status=4)
[r]   Speed 10: mode=4, speed=0, audio=10/-1, video=-1/0, pts=0
[r]   Speed 16: OK (status=4)
[r]   Speed 16: mode=4, speed=0, audio=16/-1, video=-1/0, pts=0
[r]   Speed -2147483648: OK (status=4)
[r]   Speed -2147483648: mode=4, speed=0, audio=-2147483648/-1, video=-1/0, pts=0

[x] Start pts:
[r]   Pts -1000: Failed: 80616008
[r]   Pts -1: Failed: 80616008
[r]   Pts 0: OK (status=4)
[r]   Pts 0: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Pts 1: Failed: 80616008
[r]   Pts 1000: Failed: 80616008
[r]   Pts 3003: Failed: 80616008
[r]   Pts 6006: Failed: 80616008
[r]   Pts 87087: Failed: 80616008
[r]   Pts 276276: Failed: 80616008
[r]   Pts 279278: Failed: 80616008
[r]   Pts 279279: Failed: 80616008
[r]   Pts 282282: Failed: 80616008
[r]   Pts 2147483647: Failed: 80616008

[x] Video codecs:
[r]   Video codec 00000000: OK (status=4)
[r]   Video codec 00000000: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Video codec 00000001: Failed: 80616003
[r]   Video codec 0000000e: OK (status=4)
[r]   Video codec 0000000e: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Video codec 0000000f: Failed: 80616003
[r]   Video codec 0000001e: Failed: 80616003
[r]   Video codec 8000000e: Failed: 80616003

[x] Audio codecs (no audio):
[r]   Audio codec 00000000: OK (status=4)
[r]   Audio codec 00000000: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Audio codec 00000001: OK (status=4)
[r]   Audio codec 00000001: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Audio codec 0000000e: OK (status=4)
[r]   Audio codec 0000000e: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Audio codec 0000000f: OK (status=4)
[r]   Audio codec 0000000f: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Audio codec 0000001e: OK (status=4)
[r]   Audio codec 0000001e: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Audio codec 8000000f: OK (status=4)
[r]   Audio codec 8000000f: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0

[x] Audio codecs (with audio):
[r]   Audio codec 00000000: Failed: 80616003
[r]   Audio codec 00000001: OK (status=4)
[r]   Audio codec 00000001: mode=512, speed=0, audio=1/1, video=0/0, pts=0
[r]   Audio codec 0000000e: Failed: 80616003
[r]   Audio codec 0000000f: OK (status=4)
[r]   Audio codec 0000000f: mode=512, speed=0, audio=1/1, video=0/0, pts=0
[r]   Audio codec 0000001e: Failed: 80616003
[r]   Audio codec 8000000f: Failed: 80616003

[x] Streams:
[r]   Video stream -1: Failed: 80616006
[r]   Video stream 1: Failed: 80616006
[r]   Video stream 1 (exists): OK (status=4)
[r]   Video stream 1 (exists): mode=512, speed=0, audio=1/1, video=0/0, pts=1
[r]   Video stream 2: Failed: 80616006
[r]   Audio stream -1: OK (status=4)
[r]   Audio stream -1: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Audio stream 1: OK (status=4)
[r]   Audio stream 1: mode=4, speed=0, audio=1/-1, video=-1/0, pts=0
[r]   Audio stream 1 (exists): OK (status=4)
[r]   Audio stream 1 (exists): mode=512, speed=0, audio=1/1, video=1/0, pts=0
[r]   Audio stream 2: Failed: 80616006
